dash/include/dash/algorithm/ForEach.h

```diff
@@ -14,65 +14,27 @@ namespace dash {
  * function on its local elements only.
  * To support compiler optimization, this const version is provided
  *
- * \tparam      ElementType  Type of the elements in the sequence
- * \tparam      IndexType    Parameter type expected by function to
- *                           invoke, deduced from parameter \c func
- * \complexity  O(d) + O(nl), with \c d dimensions in the global iterators'
- *              pattern and \c nl local elements within the global range
- *
- * \ingroup     DashAlgorithms
- */
-template <
-  typename ElementType,
-  class    PatternType >
-void for_each(
-  /// Iterator to the initial position in the sequence
-  const GlobIter<ElementType, PatternType> & first,
-  /// Iterator to the final position in the sequence
-  const GlobIter<ElementType, PatternType> & last,
-  /// Function to invoke on every index in the range
-  ::std::function<void(const ElementType &)> & func)
-{
-  /// Global iterators to local index range:
-  auto index_range  = dash::local_index_range(first, last);
-  auto lbegin_index = index_range.begin;
-  auto lend_index   = index_range.end;
-  auto & team       = first.pattern().team();
-  if (lbegin_index != lend_index) {
-    // Pattern from global begin iterator:
-    auto pattern      = first.pattern();
-    // Local range to native pointers:
-    auto lrange_begin = (first + pattern.global(lbegin_index)).local();
-    auto lrange_end   = lrange_begin + lend_index;
-    std::for_each(lrange_begin, lrange_end, func);
-  }
-  team.barrier();
-}
-
-/**
- * Invoke a function on every element in a range distributed by a pattern.
- * This function has the same signature as std::for_each but
- * Being a collaborative operation, each unit will invoke the given
- * function on its local elements only.
+ * \tparam      ElementType   Type of the elements in the sequence
+ * \tparam      UnaryFunction Function to invoke for each element
+ *                            in the specified range with signature
+ *                            \c void (const ElementType &)
  *
- * \tparam      ElementType  Type of the elements in the sequence
- * \tparam      IndexType    Parameter type expected by function to
- *                           invoke, deduced from parameter \c func
  * \complexity  O(d) + O(nl), with \c d dimensions in the global iterators'
  *              pattern and \c nl local elements within the global range
  *
  * \ingroup     DashAlgorithms
  */
 template <
   typename ElementType,
-  class    PatternType >
+  class    PatternType,
+  class    UnaryFunction >
 void for_each(
   /// Iterator to the initial position in the sequence
   const GlobIter<ElementType, PatternType> & first,
   /// Iterator to the final position in the sequence
   const GlobIter<ElementType, PatternType> & last,
   /// Function to invoke on every index in the range
-  ::std::function<void(ElementType &)> & func)
+  UnaryFunction                              func)
```

Comment:
`UnaryFunction` can be passed as `const &` (saves same cycles with `std::function`)